### PR TITLE
Filesystem dock folder handling

### PR DESCRIPTION
**Pull request by ghost**
Issues  #11336 and #9854 raise a point about not being able to do some common tasks from within the file system dock. 

Folders now support the following on right click:
- Move
- Rename _(The "Move or Rename" option was split, as a save as dialog doesn't make sense for folders)_
- Delete
- Copy path _(For consistency with files right click)_
- Open in explorer _(For consistency with files right click)_
- New folder

Fixes the following issues:
- Dragging a folder in the folder tree could cause broken dependencies _(possibly cause of #5684)_
- If a move or rename fails, dependencies could be updated anyway
- Files which were going to be deleted could also be listed as having broken dependencies in the dependency removal dialog.

This PR has also had to rework the rename & move code to fix bugs with/better support folders, as a result I have also spent some time tidying up and removing duplicated code in these methods. I also have reworked the DependencyRemoveDialog to properly support removing folders.

### CODE REVIEW

**Comment by MillionOstrich** [comment]
Reworked the DependencyRemovedDialog to support deleting folders. Files that were in a folder are listed under a tree item underneath the folder.  There are now some icons for the items in the tree too which makes it look a little nicer:

remove dialog

This commit also fixes a usability bug encountered during testing: Files which are being deleted may be listed as having broken dependencies if you select the file and one or more dependencies with multi-select before deleting.

Finally, because DirAccess only lets you delete empty folders (at least on windows), folder deletion happens in two stages. Files from a list generated when the dialog was opened are removed, then the remaining empty directories are cleaned up. If the user adds a file to a directory while the dialog is open the folder will not be emptied by the first step, so the deleting the directory will fail. I figured this was preferable to doing a second enumeration, as it is possible the new file has become a dependency.

**Comment by Xgor** [comment]
Was thinking of adding it in myself but I'm glad somebody else is. This better be included in 3.0 as this is probably my biggest problem with 2.1 right now. Took and checked it out and it seems to work great. The only problem I have so far is that when you drag a folder it doesn't show what folder is being dragged. I might try adding it myself but I haven't contributed yet so the Godot code is unfamiliar to me right now.

File: editor/filesystem_dock.cpp [hunk 1]
```diff
@@ -1445,97 +1504,76 @@ void FileSystemDock::drop_data_fw(const Point2 &p_point, const Variant &p_data,
...
-	//file_options->add_item(TTR("Info"),FILE_INFO);
-
-	file_options->add_separator();
+	file_options->add_item(TTR("New Folder.."), FILE_NEW_FOLDER);
 	file_options->add_item(TTR("Show In File Manager"), FILE_SHOW_IN_EXPLORER);
 
-	if (all_can_reimport && types.size() == 1) { //all can reimport and are of the same type
-
-		/*
+	/*
```
Comment (akien-mga) [inline editor/filesystem_dock.cpp]: Don't keep dead code.

**Comment by MillionOstrich** [comment]
I have decided that since I have reworked a the rename/move logic I think i should take a second look at it to fix the following two issues:

1. (Regression) moving a folder into a child folder will remap dependencies but not move the folder. There used to be a check for this and there isn't anymore.
2. More generally - if a rename or move fails the dependency remap happens anyway (this also happens in the old code). The solution to this may be to only add the remaps after the operation has been successful, then apply them after the rescan.

Whilst i'm there i also think i can do a better job at deduplicating code between move and rename.

**Comment by groud** [comment]
Some feedback from PR monday:
- We very happy for thoses changes, we hope you'll manage to fix the remaining issues :)
- It will probably need a rebase to use the new move_to_thrash function for deletion.

**Comment by akien-mga** [comment]
Sounds like 1b4f7c7 could be squashed into 728688e, and f42764b would go into 1da4149 I guess? See http://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase for docs on interactive rebase.

**Comment by MillionOstrich** [comment]
Thanks for the feedback and help so far. I am going to take a look at rebasing and shuffling around a few of these commits down tonight/tomorrow. I hope to have something that will be as follows:

- Enabling the right click menu on folders
- Move/Rename
- Delete
- Add new folder

I can squash down further after that more easily if required.
When the move_to_trash function becomes I can rebase once more if necessary.

**Comment by MillionOstrich** [comment]
Before I rebase it would be helpful if I could get some thoughts on the latest commit. I am not sure if  it is ok to use the EditorFileSystem the way I am here. 

The latest commit looks at the EditorFileSystem after the moves/renames are applied but before a call to rescan is made. This is making the assumption that the EditorFileSystem will still have the old file paths & dependencies. Doing this simplifies the code quite a bit as I don't have to keep track of this myself (which is more compicated - especially because I don't know what files will succeed in moving/renaming until actually doing it). The code I have works, but not sure if this is a reasonable requirement to put on the EditorFileSystem.

**Comment by reduz** [comment]
yeah, i guess calling to rescan later is fine

**Comment by MillionOstrich** [comment]
Rebased into the 4 commits & added a more detailed comment about the use of EditorFileSystem before rescan while I was there.

**Comment by MillionOstrich** [comment]
Rebased and updated this PR following merge of #11575.

Note that as a result of the rebase and changes to how the removal code works there is a bug introduced in the first commit which is fixed in the third. The changes to the right click in the first commit as this will let you delete ".." which is checked for in the third commit. The justification for this was that a multiselection including ".." should just skip this selected entry and try to delete the other stuff, but on reflection I think just not letting you pick some options if ".." is selected is better. Does anybody have a preference either way?

**Comment by groud** [comment]
To be honest is see no reason why '..' should be selected. Maybe that should be made impossible before ?

**Comment by MillionOstrich** [comment]
I have now made it impossible to have ".." selected (you can still right click it and double click it) and rebased this change into the first commit, fixing the issue raised in my previous comment.

**Pull request merged**